File: gtci/02-fast-and-slow-pointers/cpp/1-linked-list-cycle.cpp
#include <cassert>
#include <iostream>

using namespace std;

class ListNode {
   public:
    int val = 0;
    ListNode* next;
    ListNode(int value) {
        this->val = value;
        next = nullptr;
    }
};

class Solution {
   public:
    bool hasCycle(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;

        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;

            if (slow == fast) return true;
        }

        return false;
    }
};

int main() {
    ListNode* head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);
    head->next->next->next->next = head->next;  // create a cycle

    Solution sol;
    assert(sol.hasCycle(head) == true);

    cout << "All test cases passed." << endl;

    // Clean up memory
    for (ListNode* curr = head; curr != nullptr;) {
        ListNode* temp = curr;
        curr = curr->next;
        delete temp;
    }

    return 0;
}

File: gtci/02-fast-and-slow-pointers/python3/1-linked-list-cycle.py
class Node:
    def __init__(self, value, next=None):
        self.val = value
        self.next = next


class Solution:
    def hasCycle(self, head):
        slow = head
        fast = head

        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next

            if slow == fast:
                return True

        return False


if __name__ == "__main__":
    head = Node(1)
    head.next = Node(2)
    head.next.next = Node(3)
    head.next.next.next = Node(4)
    head.next.next.next.next = head.next  # create a cycle

    sol = Solution()
    assert sol.hasCycle(head) == True

    print("All test cases passed.")
